src/Features/Core/Portable/ConvertForEachToFor/AbstractConvertForEachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,474 @@
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ConvertForEachToFor
+{
+    internal abstract class AbstractConvertForEachToForCodeRefactoringProvider<TForEachStatement> :
+        CodeRefactoringProvider
+            where TForEachStatement : SyntaxNode
+    {
+        private const string get_Count = nameof(get_Count);
+        private const string get_Item = nameof(get_Item);
+
+        private const string Length = nameof(Array.Length);
+        private const string Count = nameof(IList.Count);
+
+        private static readonly ImmutableArray<string> s_KnownInterfaceNames =
+            ImmutableArray.Create(typeof(IList<>).FullName, typeof(IReadOnlyList<>).FullName, typeof(IList).FullName);
+
+        protected abstract string Title { get; }
+        protected abstract TForEachStatement GetForEachStatement(TextSpan selelction, SyntaxToken token);
+        protected abstract bool ValidLocation(ForEachInfo foreachInfo);
+        protected abstract (SyntaxNode start, SyntaxNode end) GetForEachBody(TForEachStatement foreachStatement);
+        protected abstract void ConvertToForStatement(
+            SemanticModel model, ForEachInfo info, SyntaxEditor editor, CancellationToken cancellationToken);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var foreachStatement = GetForEachStatement(context.Span, token);
+            if (foreachStatement == null)
+            {
+                return;
+            }
+
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var semanticFact = document.GetLanguageService<ISemanticFactsService>();
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var foreachInfo = GetForeachInfo(semanticFact, options, model, foreachStatement, cancellationToken);
+            if (foreachInfo == null)
+            {
+                return;
+            }
+
+            if (!ValidLocation(foreachInfo))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new ForEachToForCodeAction(
+                    Title,
+                    c => ConvertForeachToForAsync(document, foreachInfo, c)));
+        }
+
+        protected SyntaxToken CreateUniqueName(
```

Comment:
note certain if this helper is worth it.  consider inlining body to callers.  (your call)